### ISSUE TITLE
CA2021 false positive for `.Cast<>` from generic base class in .NET 9.0 SDK

### ISSUE THREAD

**Issue by bgrainger** [dotnet/roslyn-analyzers#7357]
### Analyzer

**Diagnostic ID**: [CA2021](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2021): Type 'GenericBase<int>' is incompatible with type 'GenericDerived' and cast attempts will throw InvalidCastException at runtime

### Analyzer source

SDK: Version 9.0.100-preview.6.24238.19

```
❯ dotnet --info
.NET SDK:
 Version:           9.0.100-preview.6.24328.19
 Commit:            ef4c241666
 Workload version:  9.0.100-manifests.bbb3781c
 MSBuild version:   17.11.0-preview-24318-05+4a45d5633
```

Also tested with NuGet Package: `<PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="9.0.0-preview.24318.1">`

### Describe the bug

NOTE: this may be a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/7153. However, since that was fixed a few months ago, I would have expected the fix to be in the latest package and latest .NET 9 Preview SDK. Opening this issue just in case it's not truly fixed.

Using `OfType<TDerived>()` or `Cast<TDerived>` on a collection of `TBase<T>` where `TDerived : TBase<T>` gives a spurious message about a cast that will fail at runtime, even though the cast is actually fine.

### Steps To Reproduce

Build the following code with .NET 9 Preview 6:

```csharp
// works as expected
new List<Base>() { new Derived() }.Cast<Derived>().ToList();

// same code but generic, fails
// CastTest\Program.cs(7,1,7,77): warning CA2021: Type 'GenericBase<int>' is incompatible with type 'GenericDerived' and cast attempts will throw InvalidCastException at runtime (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2021)
new List<GenericBase<int>>() { new GenericDerived() }.Cast<GenericDerived>().ToList();

class Base
{
}

class Derived : Base
{
}

class GenericBase<T>
{
}

class GenericDerived : GenericBase<int>
{
}
```

### Expected behavior

Compiles without warning, as it did in .NET 8.0.

### Actual behavior

Issues a spurious CA2021 warning.

**Comment by buyaa-n** [comment]
> NOTE: this may be a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/7153.

It doesn't seem fixed with https://github.com/dotnet/roslyn-analyzers/issues/7153 CC @fowl2

**Comment by bgrainger** [comment]
Still a problem with Preview 7:

```
❯ dotnet --info
.NET SDK:
 Version:           9.0.100-preview.7.24407.12
 Commit:            d672b8a045
 Workload version:  9.0.100-manifests.baed1e37
 MSBuild version:   17.12.0-preview-24374-02+48e81c6f1
```

**Comment by rjgotten** [comment]
Being hit by this with the 9.0.100 RTM SDK, even on a solution targeting .NET 8.

My particular variant is from a generic base class to a generic derived class, i.e.

```cs

new Base<int>[] { new() }.OfType<Derived<int>>()

class Base<T> { }
class Derived<T> : Base<T> {}
```

**Comment by CRidge** [comment]
We're seeing something similar when compiling our dotnet 8 solution using dotnet 9.0.100 SDK:

```
dotnet --version
9.0.100
```

We have a repo pattern where the repo fetches data from Entity Framework, and ensures the returned value is the correct type - but the signature of the repo method returns a base class:

`Task<IEnumerable<BaseEntity>> GetEntitiesAsync([predicate here]);`

When doing a .OfType<RealEntity>() on the result, we get a CA2021. 

We also get CA2021 when we call .Cast<BaseEntity>() an IEnumerable<RealEntity>. What's interesting, is that RealEntity extends a class - let's call it LocalBaseEntity, which is defined in the same project and extends BaseEntity:

`internal class LocalBaseEntity : BaseEntity { ... }`

BaseEntity is defined in a custom NuGet package we have. 

Calling Cast<LocalBaseEntity>() works just fine, even if Cast<BaseEntity> does not. Could it be an issue with a NuGet package being built with dotnet 8 SDK, but the consumer using dotnet 9 SDK?

The whole thing works just fine when building with dotnet 8 SDK.

**Comment by rjgotten** [comment]
>Could it be an issue with a NuGet package being built with dotnet 8 SDK, but the consumer using dotnet 9 SDK?

The issue is more than likely that the analyzer ships as part of the SDK.
But what code exactly shipped as part of the 9.0.100 stable SDK impossible to find out, because the 
https://github.com/dotnet/roslyn-analyzers repository does not properly tag releases. [They are a disorganized mess](https://github.com/dotnet/roslyn-analyzers/releases)...

**Comment by risperdal** [comment]
Same here

```
dotnet --info
.NET SDK:
 Version:           9.0.100
 Commit:            59db016f11
 Workload version:  9.0.100-manifests.c6f19616
 MSBuild version:   17.12.7+5b8665660
```

**Comment by JoeRobich** [comment]
Resolved by #7488

**Comment by bgrainger** [comment]
This is regressed in SDK 9.0.200:

Using `global.json`:

```json
{
  "sdk": {
    "version": "9.0.103",
    "rollForward": "disable"
  }
}
```

```
> dotnet --info
.NET SDK:
 Version:           9.0.103
 Commit:            96da45d427
 Workload version:  9.0.100-manifests.c4f6226a
 MSBuild version:   17.12.24+90b52dda6

> dotnet build
Restore complete (0.4s)
  Bug7357 succeeded (0.3s) → bin\Debug\net9.0\Bug7357.dll

Build succeeded in 0.8s
```

Using `global.json`:

```json
{
  "sdk": {
    "version": "9.0.200",
    "rollForward": "disable"
  }
}
```

```
> dotnet --info
.NET SDK:
 Version:           9.0.200
 Commit:            90e8b202f2
 Workload version:  9.0.200-manifests.c4f6226a
 MSBuild version:   17.13.8+cbc39bea8

> dotnet build
Restore complete (0.4s)
  Bug7357 succeeded with 1 warning(s) (0.2s) → bin\Debug\net9.0\Bug7357.dll
    Bug7357\Program.cs(6,1): warning CA2021: Type 'GenericBase<int>' is incompatible with type 'GenericDerived' and cast attempts will throw InvalidCastException at runtime (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2021)

Build succeeded with 1 warning(s) in 0.8s
```

**Comment by borland** [comment]
I also experienced the re-introduction of the bug that @bgrainger [references above](https://github.com/dotnet/roslyn-analyzers/issues/7357#issuecomment-2655004828) in SDK 9.0.200. Can we re-open the bug please?
